langchain_postgres/vectorstores.py

```diff
@@ -322,59 +405,125 @@ def __post_init__(
         self.create_tables_if_not_exists()
         self.create_collection()
 
-    def __del__(self) -> None:
-        if isinstance(self._engine, sqlalchemy.engine.Connection):
-            self._engine.close()
+    async def __apost_init__(
+        self,
+    ) -> None:
+        """Async initialize the store (use lazy approach)."""
+        if self._async_init:  # Warning: possible race condition
+            return
+        self._async_init = True
+
+        EmbeddingStore, CollectionStore = _get_embedding_collection_store(
+            self._embedding_length
+        )
+        self.CollectionStore = CollectionStore
+        self.EmbeddingStore = EmbeddingStore
+        if self.create_extension:
+            await self.acreate_vector_extension()
+
+        await self.acreate_tables_if_not_exists()
+        await self.acreate_collection()
 
     @property
     def embeddings(self) -> Embeddings:
         return self.embedding_function
 
     def create_vector_extension(self) -> None:
+        assert not self._async_engine, "This method must be called without async_mode"
         try:
-            with self._session_maker() as session:  # type: ignore[arg-type]
-                # The advisor lock fixes issue arising from concurrent
-                # creation of the vector extension.
-                # https://github.com/langchain-ai/langchain/issues/12933
-                # For more information see:
-                # https://www.postgresql.org/docs/16/explicit-locking.html#ADVISORY-LOCKS
-                statement = sqlalchemy.text(
-                    "BEGIN;"
-                    "SELECT pg_advisory_xact_lock(1573678846307946496);"
-                    "CREATE EXTENSION IF NOT EXISTS vector;"
-                    "COMMIT;"
-                )
-                session.execute(statement)
-                session.commit()
+            with self._engine.connect() as conn:
+                _create_vector_extension(conn)
         except Exception as e:
             raise Exception(f"Failed to create vector extension: {e}") from e
 
+    async def acreate_vector_extension(self) -> None:
+        assert self.async_mode, "This method must be called with async_mode"
+
+        async with self._async_engine.begin() as conn:
+            await conn.run_sync(_create_vector_extension)
+
     def create_tables_if_not_exists(self) -> None:
-        with self._session_maker() as session:
+        assert not self._async_engine, "This method must be called without async_mode"
```

Comment:
Could you create a session_maker and and async session maker and roll the exceptions into the appropriate session maker?

Reply:
@eyurtsev 
I'm not sure I understand the proposition.
No, it is not possible to propagate exceptions from asynchronous processing to synchronous processing.

The problem is the difference between `with` and `async with`.

Reply:
@eyurtsev, I supose this change is applies.
Note, the lazy approach is for `create_tables_if_not_exists()` too.
Impossible to guarantee in async, without a lazy approach.

Reply:
I meant doing something like in the other PR in langchain -- so we remove some of the code duplication with assertions


```python
@contextmanager
def _make_session(self)
   assert that it's sync
  with self.make_session() as session() as session:
     yield session

@asynccontextmanager
async def _amake_session(self)
   assert that it's async
  async with self.make_session() as session() as session:
     yield session
```